tests/test_task_engine_cstro.py

```diff
@@ -0,0 +1,1723 @@
+import asyncio
+import logging
+import os
+import random
+import time
+from datetime import timedelta
+from pathlib import Path
+from typing import List, Optional
+from unittest.mock import AsyncMock, MagicMock, call
+from uuid import UUID, uuid4
+
+import anyio
+import pytest
+
+from prefect import Task, flow, task
+from prefect.cache_policies import FLOW_PARAMETERS
+from prefect.client.orchestration import SyncPrefectClient
+from prefect.client.schemas.objects import StateType, TaskRun
+from prefect.context import (
+    EngineContext,
+    FlowRunContext,
+    TaskRunContext,
+    get_run_context,
+)
+from prefect.events.worker import EventsWorker
+from prefect.exceptions import CrashedRun, MissingResult
+from prefect.filesystems import LocalFileSystem
+from prefect.logging import get_run_logger
+from prefect.results import PersistedResult, ResultFactory, UnpersistedResult
+from prefect.settings import (
+    PREFECT_EXPERIMENTAL_ENABLE_CLIENT_SIDE_TASK_ORCHESTRATION,
+    PREFECT_TASK_DEFAULT_RETRIES,
+    temporary_settings,
+)
+from prefect.states import Running, State
+from prefect.task_engine import TaskRunEngine, run_task_async, run_task_sync
+from prefect.task_runners import ThreadPoolTaskRunner
+from prefect.testing.utilities import exceptions_equal
+from prefect.utilities.callables import get_call_parameters
+from prefect.utilities.engine import propose_state
+
+
+@pytest.fixture(autouse=True)
+def enable_client_side_task_run_orchestration():
+    with temporary_settings(
+        {PREFECT_EXPERIMENTAL_ENABLE_CLIENT_SIDE_TASK_ORCHESTRATION: True}
+    ):
+        yield
+
+
+async def task_run_from_last_event(
+    asserting_events_worker: EventsWorker, task_run_id: Optional[UUID] = None
+) -> TaskRun:
+    events = sorted(asserting_events_worker._client.events, key=lambda e: e.occurred)
```

Comment:
can I get away with not calling `asserting_events_worker.drain()` here? Is there a flush I should call instead? 
I want to be able to do this multiple times without stopping the `QueueService`